Alt SPIKE: New objectid (#596)
loopbackio/loopback-connector-mongodb: SPIKE.md

# Spike: Robust handling of ObjectID type for MongoDB

Ref: https://github.com/strongloop/loopback-next/issues/3456

## Table of contents

- [The problem](#the-problem)
- [Proposed solution](#proposed-solution)
- [Approaches](#approaches)
- [Developer Notes](#developer-notes)
- [Implementation Proposal](#implementation)
- [Follow-up task](#follow-up-task)

## The Problem

`ObjectID` is a native MongoDB datatype, which is used as the datatype of its primary key (_id_) field; other fields
can also be of `ObjectID` type.

> 

Reviewer: Let's add a short paragraph to explain that:

MongoDB distinguishes between id values that are a `string` and an `ObjectID`. A query using a `string` value does not match records having `ObjectID` value and vice versa. This is especially important for foreign keys, e.g. when `Order` property `customerId` is referencing an `id` of a `Customer`, then both properties must always hold `ObjectID` values. If one of the property (typically `customerId`) is stored as a `string`, then LB relations stop working.

In the current version of Juggler, the interpretation of `ObjectID` field is a sub-optimal experience supported by a complex
determination process in an already complex codebase.

Examples of bad experiences: 

> 1. Fields need not be set to `mongodb: {dataType: 'ObjectID'}` to be interpreted as `ObjectID`.

Reviewer: I find this sentence a bit difficult to understand. Are you trying to say this? 👇🏻 

```suggestion
1. Model properties must be defined as `{type: 'string', mongodb: {dataType: 'ObjectID'}}` to be interpreted as `ObjectID`.
```

Author: Nah, I am talking about the current behavior of interpreting any `ObjectID`-like string as an `ObjectID`.

2. Unless `strictObjectIDCoercion` is set to `true`, any string
that looks like an `ObjectID` will automatically get converted to `ObjectID` and you won't be able to find the object if you
tried searching the database by that string, because it is not a string any more.
3. To add to the confusion, this behavior can be configured at model and property levels.

Complexity is sometimes unavoidable, but a less complex state is always the desired state. It creates better experience for
the consumers, and makes development and maintenance a more pleasant experience for the creators.

## Proposed Solution

> "Do not interpret any property not set with `mongodb: {dataType: 'ObjectID'}` as an `ObjectID`. Use string values for `ObjectID`
> in public APIs and manage the coercion processes for the user."

Reviewer: > Do not interpret any property not set

In my experience, double negatives are difficult to understand, especially for native English speaker. Let's rephrase this in a positive way (what we want, instead what we do not want).

```suggestion
Database-specific types like `ObjectID` and `Decimal128` should be considered as an implementation detail of a database connector. Values of these types must not leak from the connector to juggler and user code, code outside the connector should see a database-agnostic type like a `string`. The connector must always convert incoming values (property values, filtering queries) from database-agnostic type to database-specific type on input, and convert database-specific types back to database-agnostic type on output.

Specifically for `ObjectID`:

- Properties that should be stored as `ObjectID` in MongoDB must be defined as `{type: 'string', mongodb: {dataType: 'ObjectID'}}`.
- There will be no auto-magic coercion from `string` to `Object` for other properties (those defined without `dataType: 'ObjectID'`).
```

That sums up the proposed solution. Users will no longer have to deal with `ObjectID` anywhere in their experience with
LoopBack/Juggler except only in one location - the model file.

## Approaches

### 1. Top level ObjectID

A top level `dataType: 'ObjectID'` was suggested, but it was discarded for the following reason:

1. If a top level `ObjectID` is supported for MongoDB, then database-specific types like `MULTIPOLYGON`, `VARCHAR`, `NCLOB` etc
may also be required to be supported. Considering the number of database-specific types and the number of databases we support,
we will end up with a really huge list.
2. Inflexible when switching connectors. Cannot do this:

```js
{
  type: 'number',
  jsonSchema: {
    format: 'int32',
  },
  mysql: {
    dataType: 'bit',
  },
  postgresql: {
    dataType: 'bit',
    length: 1,
  }
}
```

### 2. mongodb: {dataType: 'objectID'}

Setting a property to `mongodb: {dataType: 'ObjectID'}`, as it is already being done, is the better approach for marking the 

> property as an `ObjectID`.

Reviewer: Can you please mention that we are already using the same approach for data type `decimal128`?

## Proof of concept

The tests in `test/new-objectid.test.js` is a demonstration of the behavior of the new `mongodb: {dataType: 'ObjectID'}` property.

## Implementation

All changes are limited to `loopback-connetor-mongodb` only, Juggler is not affected.

- Before querying database:

>   - Convert `id` value to `ObjectID`.

Reviewer: Is it possible to use an id value that's not an `ObjectID`, e.g. a number? I vaguely remember that in such case, LoopBack 3 may be using a new property `_id` to hold the actual database id (e.g. `{_id: objectIdGeneratedByMongo, id: 123}`).

Can you please verify?

This item caught my attention because I find it a bit odd that we have to treat `id` (primary key) values differently from other properties, considering that both primary keys and other properties must be defined the same way (`mongodb: {dataType: 'ObjectID'}`).

Also if we have special treatment for primary keys, then shouldn't we also apply the same rules to foreign keys? 

Have you considered a simpler design, where we apply the same rules for all properties, regardless of whether they are a primary key or not? Is there any edge case that not work with that solution?

Author: The special treatment for primary key is the existing behavior, let me update the implementation to apply a common rule for all.

Author: > Is it possible to use an id value that's not an ObjectID, e.g. a number?

Yes, but using non-ObjectID `_id`  it comes with unexpected behaviors, so not worth using it.

Reviewer: > > Is it possible to use an id value that's not an ObjectID, e.g. a number?
> 
> Yes, but using non-ObjectID `_id` it comes with unexpected behaviors, so not worth using it.

I am not referring to non-ObjectID `_id` field. I mean a model that's using a number as the primary key on the LoopBack side.

Here is some code you can add to `lib/objectid.test.js` to better understand what's going on.

```js
  context.only('numeric id', function() {
    const User = ds.createModel(
      'UserNum',
      {
        id: {type: 'number', id: true, generated: false},
        email: {type: 'string'},
      },
      {strictObjectIDCoercion: true},
    );

    beforeEach(() => User.deleteAll());
 
    it('is supported and uses _id: ObjectID in background', async () => {
      console.log('id definition', User.definition.properties.id);

      const u = await User.create({id: 1, email: 'test@example.com'});
      console.log(u);
      console.log(u.toJSON());

      const found = await User.findById(u.id);
      console.log('found', found);
    });
  });
```

Console output:

```
id definition {
  type: [Function: Number],
  id: true,
  generated: false,
  updateOnly: false
}
{ id: 1, email: 'test@example.com' }
{ id: 1, email: 'test@example.com' }
found { email: 'test@example.com', id: 1 }
```

Interestingly enough, when I change `id` definition and remove `generated: false`, and don't provide any `id` value when creating a new user, then the connector somehow starts storing ObjectID values in the numeric `id` property.

```
id definition { type: [Function: Number], id: true, updateOnly: false }
{ email: 'test@example.com', id: 5f5b71894004d51b2a65de00 }
{ id: 5f5b71894004d51b2a65de00, email: 'test@example.com' }
found { email: 'test@example.com', id: NaN }
```

This looks like a bug to me, notice that the model instance returned by `findById` has `id: NaN`.

Author: @bajtos 

> Is it possible to use an id value that's not an ObjectID, e.g. a number? I vaguely remember that in such case, LoopBack 3 may be using a new property _id to hold the actual database id (e.g. {_id: objectIdGeneratedByMongo, id: 123}).

Yes, it possible to use an id value that's not an ObjectID, but the custom id field is not saved in the database, it is logically translated in the model from `_id` to whatever name was specified.

> when I change id definition and remove generated: false, and don't provide any id value when creating a new user, then the connector somehow starts storing ObjectID values in the numeric id property.

This happens on the master branch too. We should throw an error saying `id` is required, like how specifying the `id` for `generated: 1` throws an error.

Need some clarifications:
- If the model is defined as `{id: {id: true, type: 'string', generated: true }}`, should it be interpreted as `{id: {id: true, type: 'string', mongodb: {dataType: 'objectID'}}}`? After all, the autogenerated id will be an `ObjectID`.
- How should `{id: {id: true, type: 'number', generated: true }}` be treated? It is essentially saying `{id: {id: true, type: 'number', mongodb: {dataType: 'objectID'}}}`.

I have created a branch - https://github.com/strongloop/loopback-connector-mongodb/tree/non-objectid-id - for quickly testing the current behavior on `master`. File of interest: https://github.com/strongloop/loopback-connector-mongodb/blob/non-objectid-id/test/new-objectid.test.js.

Author: > This looks like a bug to me, notice that the model instance returned by findById has id: NaN.

Can we address this in a separate issue since the fix lies in Juggler? We will be able to make faster progress by not including changes in Juggler. From my initial look, it seems to be caused by the difference in behavior of the various DAO methods as [noted here](https://github.com/strongloop/loopback-next/issues/889#issuecomment-683887309).

Reviewer: > Yes, it possible to use an id value that's not an ObjectID, but the custom id field is not saved in the database, it is logically translated in the model from `_id` to whatever name was specified.

Interesting, I thought the model would have two properties - user-defined `id` and database-generated `_id`. Thank you for taking a look and finding out what's the actual behavior.

> If the model is defined as `{id: {id: true, type: 'string', generated: true }}`, should it be interpreted as `{id: {id: true, type: 'string', mongodb: {dataType: 'objectID'}}}`? After all, the autogenerated id will be an `ObjectID`.

One option is to silently (auto-magically) assume `mongodb: {dataType: 'objectID'}`. A better solution may be to throw a descriptive error and asks users to specify `mongodb: {dataType: 'objectID'}` in the property definition.

> * How should `{id: {id: true, type: 'number', generated: true }}` be treated? It is essentially saying `{id: {id: true, type: 'number', mongodb: {dataType: 'objectID'}}}`.

IMO, `{type: 'number', mongodb: {dataType: 'objectID'}}` is not a valid property definition, similarly for `{type: 'date', mongodb: {dataType: 'objectID'}}`, etc.

I am proposing to throw an error when such property definition is detected.

Connectors can provide `define`, `defineProperty` and `defineForeignKey` functions that are called by juggler's `DataSource` object to notify the connector about new models/properties/foreign-keys defined by the application. I think it would be best to implement these methods in MongoDB connector to validate property definitions at setup (define) time, instead of deferring the validation until the first DAO method is called.

Reviewer: > > This looks like a bug to me, notice that the model instance returned by findById has id: NaN.
> 
> Can we address this in a separate issue since the fix lies in Juggler? We will be able to make faster progress by not including changes in Juggler. From my initial look, it seems to be caused by the difference in behavior of the various DAO methods as [noted here](https://github.com/strongloop/loopback-next/issues/889#issuecomment-683887309).

Sure, that's fine. 

However, I think my proposal to reject property definitions like `{type: 'number', mongodb: {dataType: 'objectID'}}` would solve the issue and it is something we can implement inside the MongoDB connector.

Author: > I think it would be best to implement these methods in MongoDB connector to validate property definitions at setup (define) time, instead of deferring the validation until the first DAO method is called.

Let's do it in a follow up story. Let's focus specifically on `ObjectID`-to-string conversion and vice versa in this spike. If we include validation in the connector, it will further drag on the spike.

Reviewer: I feel it's important to get the developer experience right. If a user accidentally mis-configures their models, the framework should be helpful and tell them what they did wrong and how to fix it.

I am ok to work on this in a follow-up story, please make sure to include it the spike proposal. Should we start with a spike to better understand the complexities involved?

Author: >  If a user accidentally mis-configures their models, the framework should be helpful and tell them what they did wrong and how to fix it.

100% agreed, and best scoped in a separate story.

> Should we start with a spike to better understand the complexities involved?

I think so, validations are not going to be simple case of `if`-`else` when we have inferred and nested properties.

  - Walk through the data object and convert all properties defined as `mongodb: {dataType: 'ObjectID'}`.
- When receiving data from database

>   - Convert `id` value to `String`:

Reviewer: Ditto - why do we need special handling of primary keys?

Author: The special treatment for primary key is the existing behavior, let me update the implementation to apply a common rule for all.

Reviewer: I think you did not update the implementation yet, or at least I am not able to find those changes.

  - Walk through the data object and convert all properties defined as `mongodb: {dataType: 'ObjectID'}` to `String`
- Remove or refactor all helper functions and properties like `typeIsObjectId()`, `strictObjectIDCoercion` etc., related to previous behavior of `ObjectId`.

> - There are some pieces of code that can be refactored, which involves the new changes. I think we should refactor it now now, than accumulate tech debt.

Reviewer: +1 to refactor and clean up related code.

IMO, we should find a way how to unify the (more recent) implementation of `Decimal128` coercion with the pre-historic coercion of `ObjectID`. 

From what I remember, I was guiding the implementation of `Decimal128` coercion to introduce helpers that can be used for other data types like `ObjectID` too, for example `visitAllProperties` and `hasDataType`. I think we should check how far we can get with reusing those helpers for `ObjectID`, and them make the necessary improvements to fix them to support any edge cases that don't work.

Also eventually, I would like to make some of these helpers database-agnostic and move them to `loo
pback-connector`, so that we can start using them in other connectors too.

## Proposed Solution

Database-specific types like `ObjectID` and `Decimal128` should be considered as an implementation detail of a database connector.

Values of these types must not leak from the connector to Juggler and user code, code outside the connector should see a database-agnostic type like a `string`. The connector must always convert incoming values (property values, filtering queries) from database-agnostic type to database-specific type on input, and convert database-specific types back to database-agnostic type on output.

Specifically for `ObjectID`:
- Properties that should be stored as `ObjectID` in MongoDB must be defined as `{type: 'string', mongodb: {dataType: 'ObjectID'}}`.

> - There will be no auto-magic coercion from `string` to `Object` for other properties (those defined without `dataType: 'ObjectID'`).

Reviewer: ```suggestion
- There will be no auto-magic coercion from `string` to `ObjectID` for other properties (those defined without `dataType: 'ObjectID'`).
```
